src/components/sidenav/sidenav.ts

```diff
@@ -0,0 +1,353 @@
+   * Android so we use any.
+   * @param e The event.
+   * @private
+   */
+  @HostListener('transitionend', ['$event']) private onTransitionEnd_(e: Event) {
+    if (e.target == this.elementRef_.nativeElement
+        // Simpler version to check for prefixes.
+        && (e.propertyName.endsWith('transform')) {
+      this.transition_ = false;
+      if (this.opened_) {
+        if (this.openPromise_) {
+          this.openPromiseResolve_();
+        }
+        if (this.closePromise_) {
+          this.closePromiseReject_();
+        }
+
+        this.onOpen.emit(null);
+      } else {
+        if (this.closePromise_) {
+          this.closePromiseResolve_();
+        }
+        if (this.openPromise_) {
+          this.openPromiseReject_();
+        }
+
+        this.onClose.emit(null);
+      }
+
+      this.openPromise_ = null;
+      this.closePromise_ = null;
+    }
+  }
+
+  @HostBinding('class.md-sidenav-closing') private get isClosing_() {
+    return !this.opened_ && this.transition_;
+  }
+  @HostBinding('class.md-sidenav-opening') private get isOpening_() {
+    return this.opened_ && this.transition_;
+  }
+  @HostBinding('class.md-sidenav-closed') private get isClosed_() {
+    return !this.opened_ && !this.transition_;
+  }
+  @HostBinding('class.md-sidenav-opened') private get isOpened_() {
+    return this.opened_ && !this.transition_;
+  }
+  @HostBinding('class.md-sidenav-end') private get isEnd_() {
+    return this.align == 'end';
+  }
+  @HostBinding('class.md-sidenav-side') private get modeSide_() {
+    return this.mode == 'side';
+  }
+  @HostBinding('class.md-sidenav-over') private get modeOver_() {
+    return this.mode == 'over';
+  }
+  @HostBinding('class.md-sidenav-push') private get modePush_() {
+    return this.mode == 'push';
+  }
+
+  /**
+   * This is public because we need it from MdSidenavLayout, but it's undocumented and should
+   * not be used outside.
+   * @returns {number}
+   * @private
+   */
+  public get width_() {
+    if (this.elementRef_.nativeElement) {
+      return this.elementRef_.nativeElement.offsetWidth;
+    }
+    return 0;
+  }
+
+  private transition_: boolean = false;
+  private openPromise_: Promise<void>;
+  private openPromiseResolve_: () => void;
+  private openPromiseReject_: () => void;
+  private closePromise_: Promise<void>;
+  private closePromiseResolve_: () => void;
+  private closePromiseReject_: () => void;
+}
+
+
+
+/**
+ * <md-sidenav-layout> component.
+ */
+@Component({
+  selector: 'md-sidenav-layout',
+  directives: [MdSidenav],
+  templateUrl: './components/sidenav/sidenav.html',
+  styleUrls: ['./components/sidenav/sidenav.css'],
+})
+export class MdSidenavLayout implements AfterContentInit {
+  @ContentChildren(MdSidenav) private drawers_: QueryList<MdSidenav>;
+
+  get start() { return this.start_; }
+  get end() { return this.end_; }
+
+  constructor(@Optional() @Host() private dir_: Dir) {
+    if (dir_) {
```

Comment:
Could use some explanation here:
```
// If there is a `dir` attribute somewhere up the tree, we want to listen to its value so that we know to
// invert the left/right alignment of the sidenavs.
```